daemon/cmd/datapath.go

```diff
@@ -534,6 +535,68 @@ func setupStrictModeMap(lns *node.LocalNodeStore) error {
 	return nil
 }
 
+// This is a fix for the performance degradation when enabling WireGuard node
+// encryption. See (2.) in https://github.com/cilium/cilium/issues/28413#issuecomment-1898943563
+// Strictly speaking, we don't need to exectue this when WireGuard node encryption
+// is not enabled. We stay on the safe side regarding the MSS value, since
+// the RouteMTU also includes the tunneling overhead which does not affect
+// node to node communication.
+func setMSSForNodeCIDR(lns *node.LocalNodeStore, routeMTU int) {
+	go func() {
+		for {
+
+			ln, err := lns.Get(context.Background())
+			if err != nil {
+				log.WithError(err).Fatal("unable to get local node")
+			}
+			if len(ln.GetNodeInternalIPv4().String()) == 0 {
+				time.Sleep(10 * time.Second)
+				fmt.Printf("Waiting for node IP to be assigned: %s\n", ln.GetNodeInternalIPv4().String())
+				continue
+			}
+			cmd := exec.CommandContext(context.Background(), "ip", "route", "show", "match", ln.GetNodeInternalIPv4().String())
+			out, err := cmd.CombinedOutput()
+			if err != nil {
+				log.WithError(err).WithField("output", string(out)).Fatal("unable to find route")
+			}
+			lines := strings.Split(string(out), "\n")
+			nonDefaultRoutes := []string{}
+			for _, line := range lines {
+				if !strings.Contains(line, "default") && strings.Contains(line, "dev") {
+					nonDefaultRoutes = append(nonDefaultRoutes, line)
+				}
+			}
+
+			for _, route := range nonDefaultRoutes {
+				route = "route replace " + route + "advmss " + fmt.Sprintf("%d", routeMTU-40)
+				cmd := exec.CommandContext(context.Background(), "ip", strings.Split(route, " ")...)
+				fmt.Println(cmd.String())
+				if out, err := cmd.CombinedOutput(); err != nil {
+					log.WithError(err).WithField("output", string(out)).Fatal("unable to add route")
+				}
+			}
+
+			// GCP does not have any nodeCIDR specific route.
+			// Since we don't want to override the default route, we create one
+			// ourselves.
+			if len(nonDefaultRoutes) == 0 {
+				for _, nodeCIDR := range option.Config.EncryptionStrictModeNodeCIDRs {
+					gatewayIP := nodeCIDR.Addr().As4()
+					gatewayIP[3] = 1
+					gatewayIPStr := netip.AddrFrom4(gatewayIP).String()
+
+					cmd := exec.CommandContext(context.Background(), "ip", "route", "replace", nodeCIDR.String(), "via", gatewayIPStr, "advmss", fmt.Sprintf("%d", routeMTU-40))
```

Comment:
As far as I can tell, this does not change any behavior since this is executed when only the default route is found. This default route always routes via the gateway.
See the routes with this PR (the only difference is the added second to last line):
```
root@fedora:/home/cilium# ip r
default via 192.168.178.1 dev ens3 proto dhcp src 192.168.178.2 metric 1024
10.10.2.10 dev lxc696c4d1400aa proto kernel scope link
10.10.2.19 dev lxc555e298a105b proto kernel scope link
10.10.2.22 dev lxcbdaf66f8146c proto kernel scope link
10.10.2.34 dev lxc16a4c048ed81 proto kernel scope link
10.10.2.43 dev lxcc55dceb4b8a9 proto kernel scope link
10.10.2.48 dev lxc8d084a76ba8d proto kernel scope link
10.10.2.54 dev lxcfdf5ab266ea6 proto kernel scope link
10.10.2.56 dev lxcb496fef67784 proto kernel scope link
10.10.2.66 dev lxc86d58bbe292a proto kernel scope link
10.10.2.103 dev lxcfc33c0e12229 proto kernel scope link
10.10.2.109 dev lxc_health proto kernel scope link
10.10.2.113 dev lxc09ebd9afbb07 proto kernel scope link
10.10.2.127 dev lxcbf80c140d95b proto kernel scope link
10.10.2.138 dev lxcefad4319f120 proto kernel scope link
10.10.2.142 dev lxc949131d1975c proto kernel scope link
10.10.2.172 dev lxcdaaa6ed515fb proto kernel scope link
10.10.2.185 dev lxc9c5f42eeb515 proto kernel scope link
10.10.2.190 dev lxcbc0ced1fd24e proto kernel scope link
10.10.2.222 dev lxcdb6e12574f50 proto kernel scope link
10.10.2.245 dev lxcd831c75521f1 proto kernel scope link
169.254.169.254 via 192.168.178.1 dev ens3 proto dhcp src 192.168.178.2 metric 1024
192.168.178.0/24 via 192.168.178.1 dev ens3 advmss 8776
192.168.178.1 dev ens3 proto dhcp scope link src 192.168.178.2 metric 1024
```

Reply:
Oh, but if there is a gateway on the default route shouldn't we be using that instead of the first IP of the nodeCIDR in this loop iteration?

Reply:
That's a good point. Changed it to use the default gateway.